modules/overrides/user_configurable_overrides.go

```diff
@@ -0,0 +1,229 @@
+	mgr.subservicesWatcher.WatchManager(mgr.subservices)
+
+	mgr.Service = services.NewBasicService(mgr.starting, mgr.running, mgr.stopping)
+
+	return &mgr, nil
+}
+
+func (o *userConfigurableOverridesManager) starting(ctx context.Context) error {
+	if err := services.StartManagerAndAwaitHealthy(ctx, o.subservices); err != nil {
+		return errors.Wrap(err, "unable to start overrides subservices")
+	}
+
+	return o.reloadAllTenantLimits(ctx)
+}
+
+func (o *userConfigurableOverridesManager) running(ctx context.Context) error {
+	ticker := time.NewTicker(o.cfg.PollInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-ticker.C:
+			err := o.reloadAllTenantLimits(ctx)
+			if err != nil {
+				level.Error(o.logger).Log("msg", "failed to refresh user-configurable config", "err", err)
+			}
+			continue
+
+		case err := <-o.subservicesWatcher.Chan():
+			return errors.Wrap(err, "overrides subservice failed")
+		}
+	}
+}
+
+func (o *userConfigurableOverridesManager) stopping(error) error {
+	return services.StopManagerAndAwaitStopped(context.Background(), o.subservices)
+}
+
+func (o *userConfigurableOverridesManager) reloadAllTenantLimits(ctx context.Context) error {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "userConfigurableOverridesManager.reloadAllTenantLimits")
+	defer span.Finish()
+
+	traceID, _ := tracing.ExtractTraceID(ctx)
+	level.Info(o.logger).Log("msg", "reloading all tenant limits", "traceID", traceID)
+
+	// List tenants with user-configurable overrides
+	tenants, err := o.client.List(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Clean up cached tenants that have been removed from the backend
+	for cachedTenant := range o.tenantLimits {
+		if !slices.Contains(tenants, cachedTenant) {
+			o.deleteTenantLimit(cachedTenant)
+		}
+	}
+
+	// For every tenant with user-configurable overrides, download and cache them
+	for _, tenant := range tenants {
+		limits, err := o.client.Get(ctx, tenant)
+		if err != nil {
+			return errors.Wrapf(err, "failed to load tenant limits for tenant %v", tenant)
+		}
+		o.setTenantLimit(tenant, limits)
+	}
+
+	return nil
+}
+
+func (o *userConfigurableOverridesManager) getTenantLimits(userID string) (*api.UserConfigurableLimits, bool) {
+	o.mtx.RLock()
+	defer o.mtx.RUnlock()
+
+	tenantLimits, ok := o.tenantLimits[userID]
+	return tenantLimits, ok
+}
+
+func (o *userConfigurableOverridesManager) getAllTenantLimits() tenantLimits {
+	o.mtx.RLock()
+	defer o.mtx.RUnlock()
+
+	return o.tenantLimits
+}
+
+func (o *userConfigurableOverridesManager) setTenantLimit(userID string, limits *api.UserConfigurableLimits) {
+	o.mtx.Lock()
+	defer o.mtx.Unlock()
+
+	o.tenantLimits[userID] = limits
+}
+
+func (o *userConfigurableOverridesManager) deleteTenantLimit(userID string) {
+	o.mtx.Lock()
+	defer o.mtx.Unlock()
+
+	delete(o.tenantLimits, userID)
+}
+
+func (o *userConfigurableOverridesManager) Forwarders(userID string) []string {
```

Comment:
i can honestly say i've not yet seen this pattern in go. are you "overriding" the composed interface?

Reply:
It's called struct embedding. This page explains it way better than I ever could: https://eli.thegreenplace.net/2020/embedding-in-go-part-1-structs-in-structs/

By embedding `overrides.Interface` in `userConfigurableOverridesManager` our manager inherits the full implementation. And by overwriting specific methods we can inject our own overrides.
For now we only implemented `Forwarders`, the other methods are automatically handled by the embedded `Interface`.

It's definitely a complicated Go feature but we thought it would be worth it as it saves us a lot of boilerplate and potential of mistakes.

The alternative is to make `overrides.Interface` a regular field, but that would require us to implement each method of `Interface` over again (34 methods!). Every implementation would just look like this:
```go
func (o *userConfigurableOverridesManager) IngestionRateStrategy() string {
	return o.interface.IngestionRateStrategy()
}
```